packages/ra-core/src/dataProvider/useUpdate.spec.tsx

```diff
@@ -1,158 +1,301 @@
-import useUpdate from './useUpdate';
+import { useUpdate } from './useUpdate';
+import {
+    ErrorCase as ErrorCasePessimistic,
+    SuccessCase as SuccessCasePessimistic,
+} from './useUpdate.pessimistic.stories';
+import {
+    ErrorCase as ErrorCaseOptimistic,
+    SuccessCase as SuccessCaseOptimistic,
+} from './useUpdate.optimistic.stories';
+import {
+    ErrorCase as ErrorCaseUndoable,
+    SuccessCase as SuccessCaseUndoable,
+} from './useUpdate.undoable.stories';
 
 describe('useUpdate', () => {
-    it('returns a callback that can be used with update arguments', () => {
-        const dataProvider: Partial<DataProvider> = {
-            update: jest.fn(() => Promise.resolve({ data: { id: 1 } } as any)),
-        };
-        let localUpdate;
-        const Dummy = () => {
-            const [update] = useUpdate();
-            localUpdate = update;
-            return <span />;
-        };
+    describe('mutate', () => {
+        it('returns a callback that can be used with update arguments', async () => {
+            const dataProvider = {
+                update: jest.fn(() =>
+                    Promise.resolve({ data: { id: 1 } } as any)
+                ),
+            } as any;
+            let localUpdate;
+            const Dummy = () => {
+                const { mutate } = useUpdate();
+                localUpdate = mutate;
+                return <span />;
+            };
 
-        renderWithRedux(
-            // @ts-expect-error
-            <DataProviderContext.Provider value={dataProvider}>
-                <Dummy />
-            </DataProviderContext.Provider>
-        );
-        localUpdate('foo', 1, { bar: 'baz' }, { id: 1, bar: 'bar' });
-        expect(dataProvider.update).toHaveBeenCalledWith('foo', {
-            id: 1,
-            data: { bar: 'baz' },
-            previousData: { id: 1, bar: 'bar' },
-        });
-    });
-
-    it('returns a callback that can be used with mutation payload', () => {
-        const dataProvider: Partial<DataProvider> = {
-            update: jest.fn(() => Promise.resolve({ data: { id: 1 } } as any)),
-        };
-        let localUpdate;
-        const Dummy = () => {
-            const [update] = useUpdate();
-            localUpdate = update;
-            return <span />;
-        };
-
-        renderWithRedux(
-            // @ts-expect-error
-            <DataProviderContext.Provider value={dataProvider}>
-                <Dummy />
-            </DataProviderContext.Provider>
-        );
-        localUpdate({
-            type: 'update',
-            resource: 'foo',
-            payload: {
+            renderWithRedux(
+                <QueryClientProvider client={new QueryClient()}>
+                    <DataProviderContext.Provider value={dataProvider}>
+                        <Dummy />
+                    </DataProviderContext.Provider>
+                </QueryClientProvider>
+            );
+            localUpdate({
```

Comment:
Mutations created by react-query's useMutation only accept one parameter. So we can't keep the previous signature

```js
update(resource, { id, data, previousdata }) 
```

which was similar to the dataProvider function. Instead, the callback signature must be

```js
mutate({ resource, id, data, previousdata }) 
```

to avoid confusion, I also changed the hook signature the same way:

```diff
-const mutation = useUpdate(resource, { id, data, previousdata }) 
+const mutation = useUpdate({ resource, id, data, previousdata }) 
```

Reply:
I've changed my mind. I've broken the signature, but in a way that makes it look like the dataProvider one:

```diff
-const [update] = useUpdate('posts', id, data, previousDate, options);
+const [update] = useUpdate('posts', { id, data, previousDate }, options);
-update(resource, id, data, previousVersion, options);
-update(resource, { id, data, previousVersion }, options);
```